Support for ROIs with different size or resolution

ageorgou opened the pull request:
Will fix #36.

- [x] New header version for latest LabView setup
- [x] Put reading/handling of ROI information in its own class for more flexibility
- [x] Read extra columns in newer setup (resolution, FOV, ...)
- [x] Create an imaging plane for each ROI acquisition series if needed

ageorgou committed:
Add a new version for the latest LabView setup

ageorgou committed:
Split off reading of ROI table

Currently does exactly the same as before, but should help us
introduce more flexibility, particularly to customise for
different versions of the LabView setup.

ageorgou committed:
Start generalising ROI reader

ageorgou committed:
Allow creating ROI readers for newest LabView

To have full access to the header information (not used yet, but
it will be to check for variable resolution/shape), we need to
create the reader earlier on, since we don't store the header
object anywhere. This also lets us use the reader across
different methods if required (e.g when adding the acquisitions
for each ROI).

ageorgou committed:
Start differentiating for variable ROIs

Right now the Readers for newer versions don't do anything
different, but some of the structure is there to allow them to.

ageorgou commented:
This is still far from done but it's a start! It may be that subclasses for `RoiReader` are not the way to go (instead, we could create the mappings and specify behaviour through functions or other classes -- composition rather than inheritance). Since there is no immediate prospect of further LabView versions, however, we might as well not overcomplicate this right now.

The main remaining changes are:
- adding the new ROI.dat columns and their types to `RoiReaderv300`
- adding a method to `RoiReader` that takes a ROI number and an `NWBFile` (or perhaps our own `NwbFile`) and creates/returns the imaging plane corresponding to the acquisition for that ROI. For older versions, this should give the `ImagingPlane` created for the corresponding Z-plane. For the latest version, and only if variable shapes are enabled, it should create a new `ImagingPlane` based on that ROI's coordinates and resolution (thinking about this, maybe it should take a channel too, so we don't duplicate between Red and Green ROIs?) This method should be called from `read_functional_data`, when determining the imaging plane for the acquisition named `ROI_NNN_Red/Green`.

alessandrofelder committed:
add new columns

alessandrofelder committed:
function call to _imaging_section needed

alessandrofelder committed:
create simple test for ROI reader

currently fails for v300, as expected

alessandrofelder committed:
move FIXME to subclass

alessandrofelder committed:
add v300 ROI.dat

jonc125 commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -184,6 +185,8 @@ def rel(file_name):
         header_fields = self.parse_experiment_header_ini(rel('Experiment Header.ini'))
         speed_data, expt_start_time = self.read_speed_data(rel('Speed_Data/Speed data 001.txt'))
+        if expt_start_time is None:
+            expt_start_time = pd.Timestamp.now()  # FIXME get experiment start time from header if not present in speed data
```
That's an interesting one, because there seems to be some inconsistency in the data! For instance, `LabViewData2020/200225_19_09_16` has in the header
```
Experiment Date Time = "25/02/2020 18:22:17"
```
but the folder name implies a later start, as does the speed data, and all 3 differ. (Speed data starts:
```
25/02/2020	19:12:13.160587	4	0.00	332.316
```

jonc125 commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -446,17 +452,19 @@ def read_speed_data(self, file_name):
+            initial_time = speed_data.index[0] - initial_offset
+            return speed_data, initial_time
+        else:
+            return None, None  # TODO throw warning here!!
```
I'm not sure we need to warn if the file isn't there. Except that if we have neither speed data nor a modern header we need to bail out with no start time.

jonc125 commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -602,13 +610,13 @@ def rel(file_name):
                                             roi_path=roi_path,
                                             dwell_time=self.imaging_info.dwell_time / 1e6)
-        elif self.labview_version is LabViewVersions.v231:
+        elif self.labview_version is LabViewVersions.v231 or self.labview_version is LabViewVersions.v300:
```
Wouldn't just `else` be enough here?

jonc125 commented on src/silverlabnwb/rois.py:
```diff
@@ -0,0 +1,145 @@
+    """A reader for LabView version 3.0.0."""
+    def __init__(self):
+        self.column_mapping = self.base_column_mapping
+        self.column_mapping.update({
```
Won't this also update `base_column_mapping` on the abstract base class? Might need a copy on the line above?

jonc125 commented on src/silverlabnwb/rois.py:
```diff
@@ -0,0 +1,145 @@
+        assert len(roi_row_idx) == 1
+        roi_row_idx = roi_row_idx[0]
+        resolution = nwb_file.roi_data.resolution[roi_row_idx]
+        z_plane = nwb_file.nwb_file.processing['Acquired_ROIs'].get("ImageSegmentation")[plane_name].imaging_plane
```
So we're still assuming here that each ROI falls entirely within an existing Z plane?

jonc125 reviewed (commented):
Looking good to me so far!

alessandrofelder committed:
adapt to possible lack of speed data file

this can (at least currently) only happen in post-2018 versions (2.3.1
and 3.0.0) of LabView, because the latter have trial information in the
header files. In the pre-2018, the speed data is the only way a
deduction of trial times is possible, so in this case we can't do
without a speed data file (and a ValueError is raised.)

This set of changes involves transferring some calculations from the
epochs (which contain the speed data) to the trials in the file. A
side-effect of this are small signature changes, due to the 1e-9 small
difference in start and stop times between epoch and trial.

alessandrofelder committed:
update signatures

trial and epoch start_time and end_time are now consistent with each
other.

alessandrofelder committed:
clarify classname to cover all modern versions

alessandrofelder committed:
fix typo in comments

alessandrofelder committed:
wip:add functionality for variable rois

tests fail for pointing mode sample data (but not for the miniscan).
the shapes and datatypes etc all match up, but the signature values are
different. This is expected because it seems, at least for the pointing
mode, the values in the TDMS file are written as
(roi1_cycle0,...,roi10_cycle0, roi1_cycle1,...,roi10_cycle1,
roi1_cycle2,...) and we are reading the first cycles_per_trial values
for roi0. But I don't understand why the miniscan test passes?

However, this code does create a reasonable-looking variable roi file for real life
data (more tests needed in the future for this though)

The failure is related to the order of things inside the TDMS file, and
the code executed in the two inner loops (channel and roi) of
_write_roi_data. But I am struggling to figure out how one would do this
well.

Maybe we need to handle legacy and variable roi code completely
separately, although it would be more elegant if we didn't have to do
this.

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -542,18 +554,20 @@ def determine_trial_times(self):
+                    epoch_name=trial,
+                    start_time=start_time if i == 0 else start_time + 1e-9,
+                    stop_time=stop_time + 1e-9,
+                    timeseries=[speed_data_ts])
```
Can this argument be an empty list? If that doesn't cause an error, it seem simpler to change this argument (e.g. `timeseries=[speed_data_ts] if speed_data_ts is not None else []`) rather than switching to trials as the "main" interval type.

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -542,18 +554,20 @@ def determine_trial_times(self):
             # We also record exact start & end times in the trial table, since our epochs
             # correspond to trials.
-            self.nwb_file.add_trial(start_time=start_time, stop_time=stop_time)
+            self.nwb_file.add_trial(start_time=start_time if i == 0 else start_time + 1e-9, stop_time=stop_time + 1e-9)
```
Oh, I didn't remember these were different! I wonder if there was a reason that was the case.

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -711,14 +712,18 @@ def read_functional_data(self, folder_path):
             # https://github.com/NeurodataWithoutBorders/pynwb/issues/673
             for roi_num, roi_ind in self.roi_mapping[plane_name].items():
                 roi_name = 'ROI_{:03d}'.format(roi_num)
+                roi_dimensions = plane[roi_ind, 'dimensions']
```
I don't think this is used below?

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -711,14 +712,18 @@ def read_functional_data(self, folder_path):
+                if 'pixels_per_miniscan' in self.roi_mapping.keys() and 'num_lines' in self.roi_mapping.keys():
+                    all_roi_dimensions[roi_num] = [int(plane[roi_ind, 'pixels_per_miniscan']), int(plane[roi_ind, 'num_lines'])]
+                else:
+                    all_roi_dimensions[roi_num] = plane[roi_ind, 'dimensions']
```
I don't understand what the intention is here. `self.roi_mapping` maps planes to another dictionary, so I don't think `pixels_per_miniscan` will ever be a valid key.

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -985,7 +989,7 @@ def add_rois(self, roi_path):
         """
         self.log('Loading ROI locations from {}', roi_path)
-        roi_data = self.roi_reader.read_roi_table(roi_path)
+        self.roi_data = self.roi_reader.read_roi_table(roi_path)
```
What do we gain by making this an attribute instead of a local variable? Is it intended to be used in other methods?

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -115,5 +120,26 @@ def __init__(self):
+            nwb_file.add_imaging_plane(
+                name=new_plane_name,
+                description='Imaging plane for variable size ROI nr. ' + str(roi_number),
+                origin_coords=z_plane.origin_coords,
```
Could we use the real origin coordinates of the ROI here? (`x_start`, `y_start`, `z_start`)

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -115,5 +120,26 @@ def __init__(self):
+                name=new_plane_name,
+                description='Imaging plane for variable size ROI nr. ' + str(roi_number),
+                origin_coords=z_plane.origin_coords,
+                grid_spacing=z_plane.grid_spacing*resolution
```
This always trips me up! So if `resolution = 3`, does that mean that the pixels are 3 times as close, or that the distance between them is 3 times larger (than the "base" case)? This suggests the latter, right?

I guess we only need the `resolution` variable if we're adding a new plane, so it can be moved here.

ageorgou committed:
Use array for ROI sizes for easier computations

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -660,15 +674,14 @@ def read_functional_data(self, folder_path):
         # Figure out timestamps, measured in seconds
-        epoch_names = self.nwb_file.epochs[:, 'epoch_name']
-        trials = [int(s[6:]) for s in epoch_names]  # names start with 'trial_'
+        trials = [s+1 for s in self.nwb_file.trials.id.data]
```
As far as I can see, we only ever care about the number of trials/epochs here, so there's a chance to simplify this a bit!

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -798,12 +810,17 @@ def _write_roi_data(self, all_rois, num_trials, cycles_per_trial,
+                    offset_from_previous_rois = 0
                     for roi_num, data_paths in all_rois.items():
+                        roi_shape = all_roi_dimensions_pixels[roi_num]
+                        roi_ch_data = ch_data[offset_from_previous_rois:
+                                              offset_from_previous_rois + cycles_per_trial*roi_shape[0]*roi_shape[1]]
+                        roi_ch_data = roi_ch_data.reshape(np.concatenate((roi_shape, [cycles_per_trial]))[::-1])
```
Right, as you say, this isn't the right chunk of data because of how the TDMS file is laid out. We can get the correct indices, though!

ageorgou reviewed (commented):
I'm not sure I understand the change from epochs to trial as the "main" interval type. Is it the case that there will always be at least one trial, even if we don't have speed data?

alessandrofelder committed:
move reading of roi file

alessandrofelder committed:
accommodate timings.py to variable size rois

rois.py now stores copy of ROI.dat as a dataframe
and gives access to n_lines_in_roi, pixels_per_line and n_rois

calculation of n_lines_in_roi and pixels_per_line moved from timings.py to
rois.py

adapt timings.py formatting to accommodate variable size rois

timings tests inelegantly avoid using abstract factory pattern
(get_reader) and instantiate subclasses directly, but simpler
this way because we don't need to create synthetic header files.

alessandrofelder committed:
fix small bug

alessandrofelder committed:
fix isort

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -74,21 +74,44 @@ def read_roi_table(self, roi_path):
+        """
+        This method determines the number of lines and the number of pixels per line by using information
+        contained in x/y_start, x/y_stop, and angle_deg.
+        :param roi_number:
```
If we use this formatting (:+1:), we might as well explain what the parameter is! (also when overridden)

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -74,21 +74,44 @@ def read_roi_table(self, roi_path):
+            n_pixels_per_line = n_y_pixels
+        if n_lines_in_roi == 0 and n_pixels_per_line == 0:
+            n_pixels_per_line = 1
+            n_lines_in_roi = 1
```
Not particularly useful here, but just in case: you can also chain assignments together, like
```python
n_pixels_per_line = n_lines_per_roi = 1
```
(as long as you're assigning the same number)

ageorgou replied:
It would be better to use the imaging mode directly here (and probably move this at the start of the function, before computing `n_y_pixels` and `n_x_pixels`. Same in the overridden method in the subclass.

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -143,3 +166,24 @@ def get_roi_imaging_plane(self, roi_number, plane_name, nwb_file):
+        :return: a tuple containing (number of lines, number of pixels per line)
+        """
+        n_y_pixels = int(self.roi_data['num_lines'][roi_number])
+        n_x_pixels = int(self.roi_data['pixels_per_miniscan'][roi_number])
```
Hm, should we convert these columns to integer when we read the file?

ageorgou commented on src/silverlabnwb/nwb_file.py:
```diff
@@ -976,7 +979,7 @@ def read_zstack(self, zstack_folder):
         self._write()
 
-    def add_rois(self, roi_path):
+    def add_rois(self):
```
We should probably update the docstring and move the parts about the `ROI.dat` file to the `RoiReader` class, now that we no longer use the file directly here.

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -74,21 +74,44 @@ def read_roi_table(self, roi_path):
             if column not in self.type_mapping:
                 self.type_mapping[column] = np.float16
-        roi_data = pd.read_csv(
+        self.roi_data = pd.read_csv(
```
Let's define this in the constructor so we can tell if we've read the file yet or not. See `get_n_rois` for an example of why. (I thought the linter would complain!)

ageorgou commented on src/silverlabnwb/rois.py:
```diff
@@ -74,21 +74,44 @@ def read_roi_table(self, roi_path):
         }
 
+    def get_n_rois(self):
+        return len(self.roi_data['roi_index'])
```
If we haven't read the file yet and we call this method, we'll get an error because `self.roi_data` is not defined. If we initialise it in the constructor with e.g. `None`, we could do instead:
```python
return 0 if self.roi_data is None else len(self.roi_data['roi_index'])
```

(we could also drop the column and get the `len` of the whole data frame)

This gives a small bit of protection, although you could argue that we **want** to get an error if we don't read the file first...

ageorgou commented on src/silverlabnwb/timings.py:
```diff
@@ -75,16 +55,20 @@ def _read_relative_times_file(self, file_path):
-        n_lines_per_cycle = self.n_rois * self.n_lines_per_roi
-        row_increments = np.arange(self.n_pixels_per_line) * self.dwell_time
+        n_lines_per_cycle = np.sum(self.roi_reader.get_lines_pixels(roi_index)[0]
+                                   for roi_index in np.arange(0, self.n_rois))
```
Nice! The plain Python `sum` and `range` should also work here.

ageorgou commented on src/silverlabnwb/timings.py:
```diff
@@ -75,16 +55,20 @@ def _read_relative_times_file(self, file_path):
+        within_cycle_offset = 0
         for roi_index in np.arange(0, self.n_rois):
             pixel_time_offsets_by_roi[roi_index] = []
+            n_lines_in_roi, n_pixels_per_line = self.roi_reader.get_lines_pixels(roi_index)
```
If we want to avoid calling `get_lines_pixels` twice for each ROI (we already did in line 58), we could hold the results in an array. I don't imagine it's particularly expensive though.

ageorgou reviewed (commented):
Looks good! I think the main part is getting the imaging mode in the `RoiReader`, and checking the result for the variable shape data (manually and with tests).

ageorgou committed:
Don't overwrite other readers' attributes!

Without the copy, creating an instance of a RoiReaderv300 would
change the column list (and other attributes) of any existing
readers, even the ClassicalRoiReader. This isn't an issue at the
moment because we only ever use one reader for the whole runtime,
but there's no reason it should.

ageorgou committed:
Check for pointing mode directly in RoiReader

ageorgou committed:
Avoid repetition, use new columns for all v3.0.0

Split the size lookup in two steps: for pointing mode, it's shared
for all RoiReaders, but for miniscans the method is different.

Also, moved the new behaviour (using the num_lines and
pixel_per_miniscan columns) to the base v3.0.0 class, rather than
the subclass for variable shapes, since the new columns are used
in the new LabVIEW setup regardless of whether variable ROIs are
allowed (I think!)

ageorgou committed:
Cast some ROI data earlier

ageorgou committed:
Tidy up some epochs/trials things

Partly reverts 4fb96 to simplify the code a bit. Document the
difference in how we store times between trials and epochs.
Also set `trial_times` in all cases, whereas before it was only set
for newer versions. This lets us use it throughout the file.

ageorgou committed:
Use each ROI's limits in its dedicated plane

ageorgou committed:
Avoid some deprecation warnings

- Pandas will stop liking indexing with [:, np.newaxis], and we
need to get the underlying numpy array instead.
- np.sum will change behaviour when a generator is passed; we can
use the builtin sum instead.
- The syntax for accessing the channel data in nptdms has changed.
- Mapping has been in abc.collections for a while now, and the old
import will stop working in Python 3.9.

ageorgou committed:
Update datasets to int in some signatures

num_lines was changed by programmatic manipulation, without running
the tests. If the tests pass for these files, I'm confident that
the signatures for the remaining ones can be overwritten by
generating them again with the latest version of the code.

ageorgou committed:
Add a test with variable resolution

We should make a smaller version of this. I edited the signature
file manually after renaming the input directory, so there may be
some inconsistencies that still need correcting!

ageorgou committed:
Fix some things to get variable ROI import working

- It seems the line numbers and pixels can be read directly from
the corresponding column, regardless of the angle.
- We still need to add the epoch name even if there is no speed
data (should have been part of 8db9623).
- `get_roi_imaging_plane` should maybe change so that it takes a
0-based index, like the other methods (or the other way around).

ageorgou committed:
Update latest signature changes for Travis

As with some other datasets, on some platforms these are stored as
32-bit and on others as 64-bit.

alessandrofelder committed:
update signatures for some large files

using the convert_num_lines_to_int.py script on
the files where the sigs weren't update yet

alessandrofelder committed:
wip: getting roi dimensions right

alessandrofelder committed:
wip:futile attempts at accounting for angle or not

alessandrofelder committed:
cast Pixels per ROI correctly

~25000 pixels need more than an 10 bit mantissa to represent ints
correctly.

alessandrofelder committed:
wip: passes MatLab comparison...

...with Lena's data (200228_14_15_57)

alessandrofelder committed:
fix bug in add_rois, remove (possibly) unnecessary get_x_y_range functionality

alessandrofelder committed:
fix missing newline

alessandrofelder committed:
minor simplification/clarifications

alessandrofelder committed:
restore get_x_y_range because needed for pixel mask

alessandrofelder committed:
update variable-roi signature

ageorgou commented:
Latest changes look good. Making a note to check that everything works with variable-shape ROIs as it's not yet tested on Travis.

ageorgou commented:
I suggest we merge this (without creating a new release yet). Have created #80 and #81 for the remaining issues/questions.
